### PR TITLE
Feat/swarinfo 240 expose user callbacks

### PR DESCRIPTION
**Pull request by stlaperriere**
The registered callbacks can now be listed upon request. 

There is a new concept: the _CallbackArgsManifest_. Basically, this is a structure that contains a description of all the particular arguments of a given user callback. The user must declare this and provide the args to the manifest in the same order as they are used in the callback.

For void callbacks, there is no need to declare a manifest.

### CODE REVIEW

**Review by cquesnel — APPROVED** [review]

File: src/hive_mind_bridge/src/MessageHandler.cpp [hunk 1]
```diff
@@ -4,6 +4,40 @@ MessageHandler::MessageHandler() {}
...
+}
+
+MessageDTO MessageHandler::handleFunctionDescriptionRequest(
+    uint32_t requestId,
+    uint32_t msgDestinationId,
+    uint32_t msgSourceId,
+    UserCallTargetDTO sourceModule,
+    FunctionDescriptionRequestDTO functionDescriptionRequest) {
+
+    std::string name = m_callbackNames[functionDescriptionRequest.getIndex()];
```
Comment (xgroleau) [inline src/hive_mind_bridge/src/MessageHandler.cpp:25]: Maybe put a check if you get an out of bound, if it is, return a generic response with  a badrequest status

File: src/hive_mind_bridge/src/MessageHandler.cpp [hunk 2]
```diff
@@ -4,6 +4,40 @@ MessageHandler::MessageHandler() {}
...
+    UserCallbackFunctionWrapper cb = m_callbacks[name];
```
Comment (xgroleau) [inline src/hive_mind_bridge/src/MessageHandler.cpp:26]: Maybe just add a comment that m_callbackNames and m_callbacks grow together, so there is no need to check if name exist here.

**Review by xgroleau — APPROVED** [review]
LGTM, the code is pretty clean and pretty much how I imagined it when I've written the proto. Small comments on semantic and one small error case that you didn't seem to handle though

File: src/hive_mind_bridge/include/hive_mind_bridge/UserCallbackArgumentWrapper.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#ifndef HIVE_MIND_BRIDGE_USERCALLBACKARGUMENTWRAPPER_H
+#define HIVE_MIND_BRIDGE_USERCALLBACKARGUMENTWRAPPER_H
+
+#include <hivemind-host/FunctionDescriptionArgumentTypeDTO.h>
+#include <string>
+
+class UserCallbackArgumentWrapper {
```
[the remark this answers is not in the document]
Comment (stlaperriere) [reply src/hive_mind_bridge/include/hive_mind_bridge/UserCallbackArgumentWrapper.h:7]: Renamed to UserCallbackArgumentDescription.

**Pull request merged**